src/uct/ib/dc/dc_mlx5_ep.h

```diff
@@ -389,23 +395,47 @@ uct_dc_mlx5_ep_basic_init(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
                           UCS_STATS_ARG(ep->super.stats));
 }
 
-static UCS_F_ALWAYS_INLINE int
+typedef enum {
+    UCT_DC_MLX5_CAN_NOT_ALLOC, /* can't use existing or create new dcis */
+    UCT_DC_MLX5_CAN_ALLOC, /* can use existing dcis for allocation */
+    UCT_DC_MLX5_SHOULD_ALLOC /* can allocate another dci by creating another dci */
+} uct_dc_mlx5_can_alloc_res_t;
```

Comment:
seems over complication to introduce another set of return codes

Reply:
I need somehow to distinguish between those cases otherwise we will need to check the same condition twice:

```c
static UCS_F_ALWAYS_INLINE uct_dc_mlx5_can_alloc_res_t
uct_dc_mlx5_iface_dci_can_alloc(uct_dc_mlx5_iface_t *iface, uint8_t pool_index)
{
    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];

    if (ucs_likely(pool->stack_top < ucs_array_length(&pool->stack))) {
        return UCT_DC_MLX5_CAN_ALLOC;
    }

    if ((ucs_array_length(&pool->stack) == iface->tx.ndci) ||
        (pool->stack_top == iface->tx.ndci)) {
        return UCT_DC_MLX5_CAN_NOT_ALLOC;
    }

    return UCT_DC_MLX5_SHOULD_ALLOC;
}

static UCS_F_ALWAYS_INLINE uct_dc_mlx5_can_alloc_res_t
uct_dc_mlx5_iface_dci_can_alloc_or_create(uct_dc_mlx5_iface_t *iface,
                                          uint8_t pool_index)
{
    uct_dc_mlx5_can_alloc_res_t can_alloc =
            uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index);
    ucs_status_t status;
    /** Here we need to check whether to append or not, a boolean response would make me check the same condition 
         again */
    if (can_alloc != UCT_DC_MLX5_SHOULD_ALLOC) {
        return can_alloc;
    }

    status = uct_dc_mlx5_dci_pool_append_dci(iface, pool_index);
    if (status != UCS_OK) {
        return UCT_DC_MLX5_CAN_NOT_ALLOC;
    }

    return UCT_DC_MLX5_CAN_ALLOC;
}

```

Alternative:
```c
static UCS_F_ALWAYS_INLINE int
uct_dc_mlx5_iface_dci_can_alloc(uct_dc_mlx5_iface_t *iface, uint8_t pool_index)
{
    uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];

    if (ucs_likely(pool->stack_top < ucs_array_length(&pool->stack))) {
        return 1;
    }

    if ((ucs_array_length(&pool->stack) == iface->tx.ndci) ||
        (pool->stack_top == iface->tx.ndci)) {
        return 0;
    }
   return 1;
}

static UCS_F_ALWAYS_INLINE uct_dc_mlx5_can_alloc_res_t
uct_dc_mlx5_iface_dci_can_alloc_or_create(uct_dc_mlx5_iface_t *iface,
                                          uint8_t pool_index)
{
   uct_dc_mlx5_dci_pool_t *pool = &iface->tx.dci_pool[pool_index];
   int can_alloc = uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index);
    ucs_status_t status;
   
    if (can_alloc && pool->stack_top < ucs_array_length(&pool->stack)) {
        return 1;
    }

    status = uct_dc_mlx5_dci_pool_append_dci(iface, pool_index);
    if (status != UCS_OK) {
        return 0;
    }

    return 1;
}

```